File: module/properties.py
import logging
import toml
import json
import os
import pathlib

from etc import constants
from module import toml_tools
from module import files
from module import dict_tools


config_content = {}




def get_json(config):

  if not os.path.exists(config):
    return {}

  if config in config_content.keys():
    return config_content[config]

  with open(config, 'r') as f:
    try:
      config_content[config] = json.load(f)
    except Exception as e:
      config_content[config] = {}
      logging.exception(e)

    return config_content[config]



def write_json(config, content):
  config_content[config] = content
  files.writeJson(content, config)





def get_config(config):

  if not os.path.exists(config):
    return {}

  if config in config_content.keys():
    return config_content[config]

  with open(config, 'r') as f:
    try:
      config_content[config] = toml.load(f)
    except Exception as e:
      config_content[config] = {}
      logging.exception(e)

    return config_content[config]



def write_config(config, content):
  config_content[config] = content
  files.writeToml(content, config)




def get_app_properties():
  app_project_config = get_config(constants.CONFIG_TOML)
  app_user_config = get_config(constants.CONFIG_USER_TOML)

  app_config = dict(dict_tools.dict_merge(app_project_config, app_user_config))

  return app_config



def get_source_properties(config, source):

  source_config=get_config(constants.SOURCE_CONFIG_TOML)
  src_suffixes = pathlib.Path(source).suffixes
  file_extensions = "".join(src_suffixes[-2:]).removeprefix('.')

  global_settings = toml_tools.get_table_element(config, ['global', 'settings', 'general'])
  type_settings = toml_tools.get_table_element(config, ['global', 'settings', 'language']).get(file_extensions, [])

  # Override source individual settings
  if source in source_config and 'settings' in source_config[source].keys():
    global_settings.update(source_config[source]['settings'])
    logging.debug(f"{source_config[source]['settings']=}")

  global_settings.update(type_settings)
  global_settings['SOURCE_FILE_NAME'] = os.path.join(config['general']['source-dir'], source).replace('\\', '/')
  global_settings['TARGET_LIB'] = get_target_lib(source, global_settings.get('TARGET_LIB'), global_settings.get('TARGET_LIB_MAPPING'))
  global_settings['OBJ_NAME'] = pathlib.Path(pathlib.Path(source).stem).stem

  set_libl = ""
  for lib in global_settings.get('LIBL', []):
    lib = lib.replace("$(TARGET_LIB)", global_settings['TARGET_LIB'])
    if len(set_libl) > 0:
      set_libl += '; '
    set_libl += config['global']['cmds']['add-lible'].replace('$(LIB)', lib)
  global_settings['SET_LIBL'] = set_libl

  return global_settings



def get_target_lib(source, target_lib=None, lib_mapping=None):

  source_lib = source.split('/')[0].lower()

  if target_lib is not None and target_lib.lower() == '*source':
    return source_lib

  if target_lib is not None:
    return target_lib.lower()

  if lib_mapping is not None:
    for k, v in lib_mapping.items():
      k = k.lower()
      if k == source_lib:
        return v.lower()

  return source_lib

File: src/module/properties.ts
// TypeScript conversion of module/properties.py

import * as fs from 'fs';
import * as path from 'path';
import { OBIConstants } from './obi_constants';
import * as tomlTools from './toml_tools';
import * as files from './files';
import * as dictTools from './dict_tools';

const toml = require('@iarna/toml');

const configContent: Record<string, any> = {};

export function getJson(config: string): any {
    if (!fs.existsSync(config)) {
        return {};
    }

    if (config in configContent) {
        return configContent[config];
    }

    try {
        const content = fs.readFileSync(config, 'utf-8');
        configContent[config] = JSON.parse(content);
    } catch (e) {
        configContent[config] = {};
        console.error(e);
    }

    return configContent[config];
}

export function writeJson(config: string, content: any): void {
    configContent[config] = content;
    files.writeJson(content, config);
}

export function getConfig(config: string): any {
    if (!fs.existsSync(config)) {
        return {};
    }

    if (config in configContent) {
        return configContent[config];
    }

    try {
        const content = fs.readFileSync(config, 'utf-8');
        configContent[config] = toml.parse(content);
    } catch (e) {
        configContent[config] = {};
        console.error(`Error loading config file ${config}`);
        console.error(e);
    }

    return configContent[config];
}

export function writeConfig(config: string, content: any): void {
    configContent[config] = content;
    files.writeToml(content, config);
}

export function getAppProperties(): any {
    const appProjectConfig = getConfig(OBIConstants.get("CONFIG_TOML") || '.obi/config.toml');
    const appUserConfig = getConfig(OBIConstants.get("CONFIG_USER_TOML") || '.obi/config.user.toml');

    const appConfig = dictTools.dictMerge(appProjectConfig, appUserConfig);

    return appConfig;
}

export function getSourceProperties(config: any, source: string): any {
    const sourceConfig = getConfig(OBIConstants.get("SOURCE_CONFIG_TOML") || '.obi/source-config.toml');
    const srcSuffixes = path.extname(source);
    const fileExtensions = srcSuffixes.replace(/^\./, '');

    let globalSettings = tomlTools.getTableElement(config, ['global', 'settings', 'general']) || {};
    const typeSettings = tomlTools.getTableElement(config, ['global', 'settings', 'language']) || {};
    
    // Override source individual settings
    if (source in sourceConfig && 'settings' in sourceConfig[source]) {
        globalSettings = { ...globalSettings, ...sourceConfig[source]['settings'] };
    }

    if (fileExtensions in typeSettings) {
        globalSettings = { ...globalSettings, ...typeSettings[fileExtensions] };
    }

    globalSettings['SOURCE_FILE_NAME'] = path.join(config['general']['source-dir'], source).replace(/\\/g, '/');
    globalSettings['SOURCE_BASE_FILE_NAME'] = path.basename(globalSettings['SOURCE_FILE_NAME']);
    globalSettings['TARGET_LIB'] = getTargetLib(source, globalSettings['TARGET_LIB'], globalSettings['TARGET_LIB_MAPPING']);
    globalSettings['OBJ_NAME'] = path.basename(source, path.extname(source));

    globalSettings['SET_LIBL'] = getSetLiblCmd(config, globalSettings['LIBL'] || [], globalSettings['TARGET_LIB']);

    return globalSettings;
}

export function getSetLiblCmd(config: any, libl: string[], targetLib: string): string {
    let setLibl = "";
    for (const lib of libl) {
        const libReplaced = lib.replace("$(TARGET_LIB)", targetLib);
        if (setLibl.length > 0) {
            setLibl += '; ';
        }
        setLibl += config['global']['cmds']['add-lible'].replace('$(LIB)', libReplaced);
    }
    return setLibl;
}

export function getTargetLib(source: string, targetLib?: string, libMapping?: Record<string, string>): string {
    const sourceLib = source.split('/')[0].toLowerCase();

    if (targetLib && targetLib.toLowerCase() === '*source') {
        return sourceLib;
    }

    if (targetLib) {
        return targetLib.toLowerCase();
    }

    if (libMapping) {
        for (const [k, v] of Object.entries(libMapping)) {
            if (k.toLowerCase() === sourceLib) {
                return v.toLowerCase();
            }
        }
    }

    return sourceLib;
}
